providers/src/airflow/providers/http/hooks/http.py

```diff
@@ -102,49 +120,64 @@ def get_conn(self, headers: dict[Any, Any] | None = None) -> requests.Session:
-                extra.pop("allow_redirects", None)  # ignore this as only max_redirects is accepted in Session
-                session.proxies = extra.pop("proxies", extra.pop("proxy", {}))
-                session.stream = extra.pop("stream", False)
-                session.verify = extra.pop("verify", extra.pop("verify_ssl", True))
-                session.cert = extra.pop("cert", None)
-                session.max_redirects = extra.pop("max_redirects", DEFAULT_REDIRECT_LIMIT)
-                session.trust_env = extra.pop("trust_env", True)
+        self._set_base_url(connection)
+        self._set_auth(session, connection)
+        self._set_extra(session, connection)
 
-                try:
-                    session.headers.update(extra)
-                except TypeError:
-                    self.log.warning("Connection to %s has invalid extra field.", conn.host)
         if headers:
             session.headers.update(headers)
 
+        self._mount_adapters(session)
+
         return session
 
+    def _set_base_url(self, connection) -> None:
+        if connection.host and "://" in connection.host:
+            self.base_url = connection.host
+        else:
+            schema = connection.schema or "http"
+            host = connection.host or ""
+            self.base_url = f"{schema}://{host}"
+            if connection.port:
+                self.base_url += f":{connection.port}"
+
+    def _set_auth(self, session: requests.Session, connection) -> None:
+        if connection.login:
+            session.auth = self.auth_type(connection.login, connection.password)
+        elif self._auth_type:
+            session.auth = self.auth_type()
+
+    def _set_extra(self, session: requests.Session, connection) -> None:
+        if connection.extra:
+            extra = connection.extra_dejson
+            extra.pop("timeout", None)
+            extra.pop("allow_redirects", None)
+            session.proxies = extra.pop("proxies", extra.pop("proxy", {}))
+            session.stream = extra.pop("stream", False)
+            session.verify = extra.pop("verify", extra.pop("verify_ssl", True))
+            session.cert = extra.pop("cert", None)
+            session.max_redirects = extra.pop("max_redirects", DEFAULT_REDIRECT_LIMIT)
+            session.trust_env = extra.pop("trust_env", True)
+
+            try:
+                session.headers.update(extra)
+            except TypeError:
+                self.log.warning("Connection to %s has invalid extra field.", connection.host)
+
+    def _mount_adapters(self, session: requests.Session) -> None:
+        if self.adapter:
+            scheme = urlparse(self.base_url).scheme or "https"
+            session.mount(f"{scheme}://", self.adapter)
+        elif self.keep_alive_adapter:
+            session.mount("http://", self.keep_alive_adapter)
+            session.mount("https://", self.keep_alive_adapter)
```

Comment:
Why do we parse the scheme in one branch but hardcode http+https in the other?

Reply:
Thank you for your insight. The reasoning behind this implementation is to account for custom cases where the schema might be empty. However, this does seem to tie back to _set_base_url. I’ll make adjustments accordingly. Appreciate you pointing this out!